tao_compiler/mlir/disc/transforms/disc_transform_legalize_to_loop.cc

```diff
@@ -217,34 +266,51 @@ LogicalResult DiscTransformLegalizeToLoopPass::handleCpuFusionOp(
     return success();
   }
 
-  // 1, Outline the fusion to a standalone module op.
-  OwningOpRef<ModuleOp> m;
-  if (failed(outlineFusionOp(fusionOp, fusionPattern, m))) {
-    return fusionOp->emitError() << "failed to outlineFusionOp\n";
-  }
-  LLVM_DEBUG(llvm::dbgs() << "After outline fusion op:\n" << m.get() << "\n");
-
-  // 2, assign a default schedule for each pattern here.
-  PatternDescription patternDescription(fusionOp, fusionPattern, shapeAnalysis);
-  if (failed(scheduleDispatcher.dispatch(patternDescription, m.get()))) {
-    return fusionOp->emitError() << "failed to assignSchedule\n";
+  // 0, inject schedule selection logic
+  // clone the fusion op, each for one candidate schedule.
+  SmallVector<Operation*> clonedFusionOps;
+  PatternDescription pd(fusionOp, fusionPattern, shapeAnalysis);
+  if (failed(injectScheduleSelectionIR(b, pd, clonedFusionOps))) {
+    return fusionOp->emitError() << "failed to injectScheduleSelectionIR\n";
   }
-  LLVM_DEBUG(llvm::dbgs() << "After assign schedule for fusion op:\n"
+  LLVM_DEBUG(llvm::dbgs() << "After injectScheduleSelectionIR:\n"
                           << m.get() << "\n");
 
-  // 3, Build a nested pass pipeline to legalize the outlined fusion op.
-  if (failed(runTransformPipeline(m.get()))) {
-    return fusionOp->emitError() << "failed to run runTransformPipeline\n";
-  }
-  LLVM_DEBUG(llvm::dbgs() << "After run transform pipeline:\n"
-                          << m.get() << "\n");
+  for (auto fusion : clonedFusionOps) {
+    b.setInsertionPoint(fusion);
+    auto fusionOp = cast<lmhlo::FusionOp>(fusion);
+    FusionPattern fusionPattern(fusionOp, &shapeAnalysis);
+    // 1, Outline the fusion to a standalone module op.
+    OwningOpRef<ModuleOp> m;
+    if (failed(outlineFusionOp(fusionOp, fusionPattern, m))) {
+      return fusionOp->emitError() << "failed to outlineFusionOp\n";
+    }
+    LLVM_DEBUG(llvm::dbgs() << "After outline fusion op:\n" << m.get() << "\n");
 
-  // 4, Inline the lowered IR into the orignal module.
-  if (failed(inlineTransformedModule(b, fusion, fusionPattern, m.get()))) {
-    return fusion->emitError() << "failed to inlineTransformedModule\n";
+    // 2, assign a default schedule for each pattern here.
+    PatternDescription patternDescription(fusionOp, fusionPattern,
+                                          shapeAnalysis);
+    if (failed(scheduleDispatcher.dispatch(patternDescription, m.get()))) {
```

Comment:
nit: Would this introduce any inconsistency with scheduleFactory selected in `injectScheduleSelectionIR`? Could it get solved if `injectScheduleSelectionIR` return corresponding candidate schedule factory with each cloned fusion op?

Reply:
it will not introduce any inconsistency. After `injectScheduleSelectionIR `, each cloned fusion is tagged with tags from the selected schedule factory. The unique id of a schedule factory is also one tag of the factory, thus there will be only one candidate.